### DISCUSSION TITLE
How to use the ModelSignature.from_dict() function

### DISCUSSION THREAD

**Discussion by dragonfly06** [mlflow/mlflow#12614] [Q&A] [answered]
I'm trying to create a ModelSignature in MLflow using [ModelSignature.from_dict()](https://mlflow.org/docs/latest/python_api/mlflow.models.html#mlflow.models.ModelSignature) function. This is my code:

```python
from mlflow.models import ModelSignature

signature_dict = {
    "inputs": '{"messages": [{"content": "string","role": "string"}]}',
    "outputs": """{
        "question": "string",
        "history": [{"content": "string", "role": "string"}],
        "context": [
            {"metadata": {"url": "string", "id": 0.0}, "page_content": "string"}
        ],
        "answer": "string"
    }""",
    "params": "",
}

model_signature = ModelSignature.from_dict(signature_dict)
```

It throws the following error: `TypeError: string indices must be integers`.
Apparently the `signature_dict` is not set up correctly, but I can't figure out what's wrong. I couldn't find a code example for this.

**Comment by B-Step62** [comment, marked as answer]
@dragonfly06 The `ModelSignature.from_dict()` accepts the dictionary representation of the signature, which is different from what you passed in.

You can see how it is constructed by checking the output of `to_dict()` method of a valid signature. For example:
```
from mlflow.models import ModelSignature
from mlflow.types import DataType
from mlflow.types.schema import *

signature = ModelSignature(
    inputs=Schema([ColSpec(DataType.boolean), ColSpec(DataType.binary)]),
    outputs=Schema(
        [ColSpec(name=None, type=DataType.double), ColSpec(name=None, type=DataType.double)]
    ),
)

print(signature1.to_dict())
> {'inputs': '[{"type": "boolean", "required": true}, {"type": "binary", "required": true}]',
 'outputs': '[{"type": "double", "required": true}, {"type": "double", "required": true}]',
 'params': None}
```

Therefore, the error could be fixed by changing the dictionary format to this, however, this method is indeed not intended to be used publicly. I would rather recommend using the easier construction method such as:

1. **Infer from an example payload** - If you have some input data sample, you can use `mlflow.models.infer_signature()` function to let MLflow construct the signature for you.
2. **Use the constructor** - Instead of passing dictionary, you can use the constructor like `ModelSignature(inputs=Schema(...), outputs=Schema(...), params=None)`. This is an official way of manually creating signature.

Please refer to the [signature documentation](https://mlflow.org/docs/latest/model/signatures.html#column-based-signature-example) for more details about above two methods.
